src/dataset.rs

```diff
@@ -1,60 +1,112 @@
-use crate::gdal_major_object::MajorObject;
-use crate::metadata::Metadata;
-use crate::raster::driver::_register_drivers;
-use crate::raster::types::GdalType;
-use crate::raster::{Driver, RasterBand, GDALDataType};
-use crate::utils::{_last_cpl_err, _last_null_pointer_err, _string};
-use gdal_sys::{self, CPLErr, GDALAccess, GDALDatasetH, GDALMajorObjectH};
+use std::{ffi::CString, path::Path, ptr, sync::Once};
+
+use crate::{
+    gdal_major_object::MajorObject, raster::RasterBand, spatial_ref::SpatialRef,
+    utils::_last_cpl_err, utils::_last_null_pointer_err, vector::Layer, Driver,
+};
+use crate::{metadata::Metadata, utils::_string};
+use gdal_sys::{
+    self, CPLErr, GDALAccess, GDALDatasetH, GDALMajorObjectH, OGRLayerH, OGRwkbGeometryType,
+};
 use libc::{c_double, c_int};
-use std::ffi::CString;
-use std::path::Path;
-use std::ptr::null_mut;
-
-#[cfg(feature = "ndarray")]
-use ndarray::Array2;
+use ptr::null_mut;
 
 use crate::errors::*;
 
 pub type GeoTransform = [c_double; 6];
+static START: Once = Once::new();
 
 pub struct Dataset {
     c_dataset: GDALDatasetH,
 }
 
-impl MajorObject for Dataset {
-    unsafe fn gdal_object_ptr(&self) -> GDALMajorObjectH {
-        self.c_dataset
+pub fn _register_drivers() {
+    unsafe {
+        START.call_once(|| {
+            gdal_sys::GDALAllRegister();
+        });
     }
 }
 
-impl Metadata for Dataset {}
+impl Dataset {
+    pub unsafe fn from_c_dataset(c_dataset: GDALDatasetH) -> Dataset {
+        Dataset { c_dataset }
+    }
 
-impl Drop for Dataset {
-    fn drop(&mut self) {
-        unsafe {
-            gdal_sys::GDALClose(self.c_dataset);
-        }
+    pub unsafe fn c_dataset(&self) -> GDALDatasetH {
+        self.c_dataset
     }
-}
 
-impl Dataset {
     pub fn open(path: &Path) -> Result<Dataset> {
+        Self::open_ex(path, None, None, None, None)
+    }
+
+    pub fn open_ex(
+        path: &Path,
+        open_flags: Option<u32>,
+        _allowed_drivers: Option<&str>, // TODO: use parameters
+        _open_options: Option<&str>,
+        _sibling_files: Option<&str>,
+    ) -> Result<Dataset> {
         _register_drivers();
         let filename = path.to_string_lossy();
         let c_filename = CString::new(filename.as_ref())?;
-        let c_dataset = unsafe { gdal_sys::GDALOpen(c_filename.as_ptr(), GDALAccess::GA_ReadOnly) };
+        let c_open_flags = open_flags.unwrap_or(GDALAccess::GA_ReadOnly); // This defaults to GdalAccess::GA_ReadOnly
+
+        let c_dataset = unsafe {
+            gdal_sys::GDALOpenEx(
+                c_filename.as_ptr(),
+                c_open_flags,
+                ptr::null(),
+                ptr::null(),
```

Comment:
`papszOpenOptions` aka `_open_options` would be the most important improvement I guess